### PR TITLE
stub: avoid race in onHalfClose server StreamObserver

### PR DESCRIPTION
**Pull request by carl-mastrangelo**

### CODE REVIEW

**Comment by zhangkun83** [comment]
LGTM

File: stub/src/main/java/io/grpc/stub/ServerCalls.java [hunk 1]
```diff
@@ -171,7 +172,7 @@ public void onHalfClose() {
         method.invoke(request, responseObserver);
         request = null;
         responseObserver.freeze();
-        if (call.isReady()) {
+        if (!wasReady) {
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: This is inverted. It should be `if (wasReady)`

Comment (carl-mastrangelo) [reply stub/src/main/java/io/grpc/stub/ServerCalls.java]: Oh, I thought you said onReady should only be called once, and it was which ever method (onHalfClose or onReady) was invoked first.  Is that not correct?

Comment (ejona86) [reply stub/src/main/java/io/grpc/stub/ServerCalls.java]: Before `onHalfClose()`, `onReadyHandler == null`. So if there was an onReady callback, the code "ignored" it. The onReady() callback will generally happen before onHalfClose(). This check was to "replay" that missed onReady call.

Comment (carl-mastrangelo) [reply stub/src/main/java/io/grpc/stub/ServerCalls.java]: I see, updated.

**Pull request merged**